presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,174 @@
+ * merged into a single WindowNode.
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `-- ProjectNode(...)
+ *             `--WindowNode(Specification: A, Functions: [avg(something)])
+ *                `--...
+ */
+public class MergeIdenticalWindows
+        implements PlanOptimizer
+{
+    public PlanNode optimize(PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new Context());
+    }
+
+    private static class Context
+    {
+        private Map<WindowNode.Specification, ImmutableMap.Builder<Symbol, FunctionCall>> functions =
+                new HashMap<>();
+        private Map<WindowNode.Specification, ImmutableMap.Builder<Symbol, Signature>> signatures =
+                new HashMap<>();
+
+        private boolean collectFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+
+            // Either both maps must contain the specification, or neither.
+            checkState(functions.containsKey(key) == signatures.containsKey(key),
+                    "Merge context is in an inconsistent state.");
+
+            ImmutableMap.Builder<Symbol, FunctionCall> functionsBuilder = functions.get(key);
+            if (functionsBuilder == null) {
+                functionsBuilder = new ImmutableMap.Builder<>();
+            }
+
+            functionsBuilder.putAll(window.getWindowFunctions());
+            functions.put(key, functionsBuilder);
+
+            ImmutableMap.Builder<Symbol, Signature> signaturesBuilder = signatures.get(key);
+            if (signaturesBuilder == null) {
+                signaturesBuilder = new ImmutableMap.Builder<>();
+            }
+            signaturesBuilder.putAll(window.getSignatures());
+            return signatures.put(key, signaturesBuilder) == null;
+        }
+
+        private static class FunctionPair<F, S>
+        {
+            final Map<Symbol, FunctionCall> windowFunctions;
+            final Map<Symbol, Signature> signatures;
+
+            private FunctionPair(
+                    Map<Symbol, FunctionCall> windowFunctions,
+                    Map<Symbol, Signature> signatures)
+            {
+                this.windowFunctions = windowFunctions;
+                this.signatures = signatures;
+            }
+        }
+
+        private FunctionPair getCollectedFunctions(WindowNode window)
+        {
+            WindowNode.Specification key = window.getSpecification();
+
+            ImmutableMap.Builder<Symbol, FunctionCall> functionBuilder = functions.remove(key);
+            ImmutableMap.Builder<Symbol, Signature> signatureBuilder = signatures.remove(key);
```

Comment:
why do you remove them? Maybe instead of removing you can not add if a function or signature already exists in collection. In such case you collect all information at the end of `Rewrite`, IMO thanks to that it will be easier to debug.